Document retired node replacement in ES (#4326)
WordPress/openverse: documentation/meta/maintenance/elasticsearch_cluster.md

If a Data refresh is scheduled during the timeframe of this procedure, it is
best to pause it for performance reasons.

### Steps

1. Use the Instance ID provided in the retirement notification to find the
   "Private IPv4 address" of the instance. This value is the "Name" of the
   ElasticSearch node.
1. Identify where the node is stored in our Terraform state. This will be useful
   for manual state changes later in the process. Check each node until you find
   the matching private IP in the returned state by running
   `j tf <env> state show "module.<env>-elasticsearch-8-8-2.aws_instance.datanodes[0]"`,
   `j tf <env> state show "module.<env>-elasticsearch-8-8-2.aws_instance.datanodes[1]"`,
   and so on for each index. Replace the "<env>" with the correct values. Record
   the final `module.<env>-elasticsearch-8-8-2.aws_instance.datanodes[x]` index.
1. In Terraform, increase the `data_node_count` for the relevant Elasticsearch
   module by one. Here is an
   [example commit](https://github.com/WordPress/openverse-infrastructure/pull/894/commits/4a827b786b1460aa89931d474db119d835784727)
   with this change in production. Apply this change and wait for a new instance

>    to be provisioned and connected to the cluster.

Author: ```suggestion
   to be provisioned and connected to the cluster.
 2. Initialize and configure the new instance with 
    `just ansible/playbook <env> elasticsearch/sync_config.yml -e apply=true`.
```

determining whether it's possible to test it locally. Otherwise, follow the
steps for deploying a new cluster and switching the live services over to it.

## Node Replacements

Occasionally, the underlying EC2 instance for one of our nodes will undergo a
scheduled decomission ("retirement") and need to be replaced beforehand.

This procedure describes the process of replacing a single node, but could be
extrapolated to multiple nodes with relative ease.

> If a Data refresh is scheduled during the timeframe of this procedure, it is

Reviewer: Nit:
```suggestion
If a data refresh is scheduled during the timeframe of this procedure, it is
```

best to pause it for performance reasons.

### Steps

1. Use the Instance ID provided in the retirement notification to find the
   "Private IPv4 address" of the instance. This value is the "Name" of the
   ElasticSearch node.
1. Identify where the node is stored in our Terraform state. This will be useful
   for manual state changes later in the process. Check each node until you find
   the matching private IP in the returned state by running
   `j tf <env> state show "module.<env>-elasticsearch-8-8-2.aws_instance.datanodes[0]"`,
   `j tf <env> state show "module.<env>-elasticsearch-8-8-2.aws_instance.datanodes[1]"`,

>    and so on for each index. Replace the "<env>" with the correct values. Record

Reviewer: This apparently doesn't render correctly in the preview 😅 
```suggestion
   and so on for each index. Replace the `<env>` with the correct values. Record
```

>    the matching private IP in the returned state by running
>    `j tf <env> state show "module.<env>-elasticsearch-8-8-2.aws_instance.datanodes[0]"`,
>    `j tf <env> state show "module.<env>-elasticsearch-8-8-2.aws_instance.datanodes[1]"`,
>    and so on for each index. Replace the "<env>" with the correct values. Record

Reviewer 2: It's better to avoid using aliases that are so short in documentation and to make this a code block instead of inline for easier copying.

````suggestion
   the matching private IP in the returned state by running
   ```
   just tf <env> state show "module.<env>-elasticsearch-8-8-2.aws_instance.datanodes[0]"
   just tf <env> state show "module.<env>-elasticsearch-8-8-2.aws_instance.datanodes[1]"
   ```
   and so on for each index. Replace the "<env>" with the correct values. Record
````

Also, even avoid using `just` here since it can be messy with the argument and the `""` sometimes. Using Terraform directly has been suggested in the past, but choose whatever you prefer.

````suggestion
   the matching private IP in the returned state by running
   ```
   cd <env>
   terraform state show "module.<env>-elasticsearch-8-8-2.aws_instance.datanodes[0]"
   terraform state show "module.<env>-elasticsearch-8-8-2.aws_instance.datanodes[1]"
   ```
   and so on for each index. Replace the "<env>" with the correct values. Record
````

Reviewer 3: We can (and should) fix the issue with just and quoting by passing arguments through the just `quote` function.

I find `just tf` to be tedious and sometimes unpredictable, and have recommended against using it because of my own confusion with its behaviour. It only sometimes does what I intended, and that frightens me with Terraform, especially direct state manipulations.

   the final `module.<env>-elasticsearch-8-8-2.aws_instance.datanodes[x]` index.
1. In Terraform, increase the `data_node_count` for the relevant Elasticsearch
   module by one. Here is an
   [example commit](https://github.com/WordPress/openverse-infrastructure/pull/894/commits/4a827b786b1460aa89931d474db119d835784727)
   with this change in production. Apply this change and wait for a new instance
   to be provisioned and connected to the cluster.
1. Initialize and configure the new instance with
   `just ansible/playbook <env> elasticsearch/sync_config.yml -e apply=true`.
1. Use `just jh es {staging|production}` to connect to the cluster, and send
   `{ "transient":{ "cluster.routing.allocation.exclude._ip": "<IP_ADDRESS>" } }`
   to the `/_cluster/settings` endpoint (in a GUI like Elasticvue or via `curl`)
   to deallocate shards from the bad node. Be sure to replace `<IP_ADDRESS>`
   with the private IPv4 address identified in step 1.

> 1. Let the cluster relocate shards to the new node and from the retired node.
>    Wait for the cluster health to return to green.

Reviewer: Wording nit:
```suggestion
1. The cluster will now relocate shards to the new node and from the retired node.
   Wait for the cluster health to return to green.
```
